icaruscode/Filters/FilterCRTPMTMatching_module.cc

```diff
@@ -0,0 +1,618 @@
+////////////////////////////////////////////////////////////////////////
+// Class:       FilterCRTPMTMatching
+// Plugin Type: analyzer (Unknown Unknown)
+// File:        FilterCRTPMTMatching_module.cc
+//
+// Generated at Thu Feb 17 13:31:00 2022 by Biswaranjan Behera using cetskelgen
+// from  version .
+////////////////////////////////////////////////////////////////////////
+
+#include "art/Framework/Core/EDFilter.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "canvas/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "art_root_io/TFileService.h"
+#include "canvas/Persistency/Common/FindManyP.h"
+
+// LArSoft includes
+#include "larcore/CoreUtils/ServiceUtil.h"
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+#include "larcorealg/Geometry/GeometryCore.h"
+#include "larcore/Geometry/Geometry.h"
+
+// Data product includes
+#include "lardataobj/RecoBase/OpHit.h"
+#include "lardataobj/RecoBase/OpFlash.h"
+#include "sbnobj/Common/CRT/CRTHit.hh"
+#include "icaruscode/CRT/CRTUtils/CRTCommonUtils.h"
+#include "sbnobj/Common/Trigger/ExtraTriggerInfo.h"
+
+// C++ includes
+#include <vector>
+#include <map>
+#include <numeric>
+
+// ROOT includes
+#include "TTree.h"
+#include "TVector3.h"
+
+namespace icarus
+{
+    namespace crt
+    {
+        class FilterCRTPMTMatching;
+        struct CRTPMT
+        {
+            double tof;
+            double distance;
+            art::Ptr<sbn::crt::CRTHit> CRTHit;
+        };
+        struct MatchedCRT
+        {
+            // vector of pairs where first is the matched Time of Flight and second is the matched CRTHit
+            std::vector<CRTPMT> entering;
+            std::vector<CRTPMT> exiting;
+            int matchType;
+        };
+        struct TrajPoint
+        {
+            double X;
+            double Y;
+            double Z;
+            double T;
+        };
+    }
+}
+
+using namespace icarus::crt;
+
+bool flashInTime(double const &flashTime, int gateType, double gateDiff, double flashTimeCut)
+{
+    std::map<unsigned int, double> gateLength{
+        {1, 2.2},  // BNB
+        {2, 10.1}, // NuMI
+        {3, 2.2},  // BNB Offbeam
+        {4, 10.1}, // NuMI Offbeam
+    };
```

Comment:
I like your suggestion. I did not know it was already included. The module does not have an art::Run object, but only art::Event (and I don't think I can access TriggerConfiguration from art:Event, right?). Do you have some suggestions on how to do this?

Reply:
For the record: I usually read the data product at the entry or each run, by having a `icarus::TriggerConfiguration fTrigConfig;` data member in the class, and then hooking
```cpp
void beginRun(art::Run& run) override {
  fTrigConfig = run.getProduct<icarus::TriggerConfiguration>(fTriggerTag);
}
```
or something like that.
It appears that `art::Event::getRun()` can also give access to `art::Run`, so that this reading can be done locally on each event. It is probably slightly less efficient, but has also advantages.